### ISSUE TITLE
Intermittent black screen in WebRTC when connected through an external network after some time

### ISSUE THREAD

**Issue by plia7** [pikvm/pikvm#1084]
**Describe the bug**
Intermittent black screen in WebRTC when connected through an external network after some time.

**To Reproduce**
Steps to reproduce the behavior, like:
1. Connect to pikvm from an external network using tailscale installed on the external network connecting machine.
2. Stay connected and use it for few hours.

**Expected behavior**
Being able to stay connected and use it for few hours without any interruptions assuming stable and fast internet connection on both ends.

**Actual behavior**
Intermittent black screen in WebRTC starts appearing after some time (could be after minutes or hours) where it works fine until that point, then prior to the black screen it might appear like the mouse movement is slower to react (although no observed internet quality reduction on both ends). Then when the black screen appears for few seconds or minutes, it shows like an orange sd memory card icon (all other icons don't have a color), and in the window caption it shows 0 fps:

<img width="243" alt="window caption during black screen" src="https://github.com/pikvm/pikvm/assets/134864072/d4730a54-8bc6-4235-9462-918fb9bbdfcc">

These are my runtime settings:

<img width="394" alt="runtime settings" src="https://github.com/pikvm/pikvm/assets/134864072/f37a3cde-90f9-452b-bbb7-cff5d9669ed3">

When opening the log, it shows:

"23-08-09 21:39:00 kvmd-janus.service] --- kvmd.apps.janus.runner            INFO --- => [WARN] [497280146060660] Waiting for candidates-done callback... (slow gathering, are you using STUN or TURN for Janus too, instead of just for users? Consider enabling full-trickle instead)
[2023-08-09 21:39:01 kvmd-janus.service] --- kvmd.apps.janus.runner            INFO --- => [WARN] [497280146060660] Failed to add some remote candidates (added 0, expected 1)"

As well more detailed log for when the issue happens:

[[2023-08-11 001109 kvmd.service] --.txt](https://github.com/pikvm/pikvm/files/12321042/2023-08-11.001109.kvmd.service.--.txt)

After sometime it seem to restore the screen, but the mouse movement and general feel is that the connection is not as stable, and it keeps going back to black screen again after few seconds so it's not very useable at this point.

Note: If I switch to mjpeg, it seem to show the screen then, but doesn't seem like the mouse/input is very reactive in that mode. Trying to click stop stream and start stream while in webrtc doesn't seem to help.

**Possible workaround although not always consistent**
It seems like if I open an incognito browser for same pikvm or the second pikvm non-incognito web page for the same host pc it might work better for some time and the black screen issue will be gone, but after some time, it starts creeping in into the incognito or second pikvm session again so I have to keep altering between pikvm 1, incognito and pikvm 2 to "escape" the issue but this is not an ideal scenario and I rather not have any interruptions to begin with, just like when I was using it in my local network without this issue.

**Screenshots**

**Desktop (please complete the following information):**
 Windows 11
Google Chrome Version 114.0.5735.199 (Official Build) (64-bit)
or
Microsoft Edge Version 115.0.1901.183 (Official build) (64-bit)
 - VNC client (if used)

**PiKVM info:**
  - Raspberry Pi board version [e.g. RPi 4]
 PiKVM v3 Pre-Assembled
  - PiKVM platform [e.g. v2-hdmi]
  - Video capture type [e.g. CSI bridge]
  - KVMD version: `pacman -Q | grep kvmd`
kvmd 3.236-1
kvmd-fan 0.26-1
kvmd-oled 0.24-1
kvmd-platform-v3-hdmi-rpi4 3.238-1
kvmd-webterm 0.43-1
  - uStreamer version: `pacman -Q | grep ustreamer`
  ustreamer 5.41-1
  - Linux kernel: `uname -a`
 Linux pikvm 5.15.68-3-rpi-ARCH #1 SMP Mon Oct 31 20:56:54 MSK 2022 armv7l GNU/Linux

**Additional context**
This is the same system setup I used to have in my local network with the connecting pc and host pc - and I never encountered this issue when everything was in the local network. I only started to encounter this issue when the connecting pc is on an external network (I haven't changed any settings from that "local network working setup").

Here is the main "trademark" of the issue, when it blinks it shows this prompt:

<img width="1188" alt="image" src="https://github.com/pikvm/pikvm/assets/134864072/f534c0a0-3d70-4bfa-89f7-5cd5e54a1039">

"H.264 - 1920x1080 / Invalid PeerConnection / 0 fps dynamic"
As you can see the sd memory card icon is still showing in green color (not yellow)

**Comment by mdevaev** [comment]
Please update to the latest version and try to reproduce it again.

**Comment by plia7** [comment]
> Please update to the latest version and try to reproduce it again.

I updated to kvmd-platform-v3-hdmi-rpi4 3.244-1.
I'll keep you posted if I still encounter this issue with the latest version.

**Comment by plia7** [comment]
> Please update to the latest version and try to reproduce it again.

@mdevaev The black screen blinking continues in the latest version. I wouldn't say it's as rare as you would expect, but happens more frequently like every x amount of minutes, although it doesn't seem to last long this time around, it's still a frustrating experience overall.

Here are the logs:

[logs 8.14.2023.txt](https://github.com/pikvm/pikvm/files/12332449/logs.8.14.2023.txt)

It happened like 2-3 times in the last 45 minutes session when I was connected. Although it doesn't seem to persist as long like before.

The logs mention things like (which I also mentioned above originally):

[2023-08-14 10:31:59 kvmd-janus.service] --- kvmd.apps.janus.runner            INFO --- => [ERR] [ice.c:janus_ice_cb_nice_recv:3078] [1829142975794202]     SRTCP unprotect error: srtp_err_status_replay_fail (len=140-->140)

And:

[2023-08-14 10:07:17 kvmd-janus.service] --- kvmd.apps.janus.runner            INFO --- => [WARN] [7136366473408474] Failed to add some remote candidates (added 0, expected 1)

Internet is stable and fast on both connecting pc (using wifi) and host pc (host pc is using wired ethernet and as well as the pikvm):

Speed test in connecting pc:

![image](https://github.com/pikvm/pikvm/assets/134864072/3761e101-bf7e-4aa5-b019-df72d410af9f)

Speed test in host pc:

![image](https://github.com/pikvm/pikvm/assets/134864072/e0ea5b04-9547-414b-8a7b-9f1e2ca2987e)

Thank you.

**Comment by plia7** [comment]
> Please update to the latest version and try to reproduce it again.

@mdevaev Just had some strong encounter with the issue, it went to became slow again in conjunction with the black screen: 

<img width="1244" alt="image" src="https://github.com/pikvm/pikvm/assets/134864072/579c83bf-629b-4c61-ac4e-727e6e4c21ec">

And the black screen seem to persist longer this time around which makes it quite unusable and unreliable when it happens and keeps happening during this session unless I go incognito or start another pikvm session, here are the logs:

[logs 8.14.2023 take 2.txt](https://github.com/pikvm/pikvm/files/12338593/logs.8.14.2023.take.2.txt)

Internet is still stable and fast on both connecting pc (using wifi) and host pc (host pc is using wired ethernet and as well as the pikvm):

Speed test in connecting pc:

![image](https://github.com/pikvm/pikvm/assets/134864072/bf262039-b060-4495-9673-0e2a7716a7ef)

Speed test in host pc:

![image](https://github.com/pikvm/pikvm/assets/134864072/a1d919f5-fe1a-444a-a996-116e1cc6eeea)

Thank you.

**Comment by mdevaev** [comment]
Note: https://github.com/meetecho/janus-gateway/blob/0.x/html/janus.js#L1949

**Comment by plia7** [comment]
@mdevaev As discussed, here are the triangles you requested:

![image](https://github.com/pikvm/pikvm/assets/134864072/9e75795a-1edf-4f4e-a444-74d726e4c282)

**Comment by plia7** [comment]
> I need to add myself to being affected by this but in my case I'm in a completely internal network. I require webrtc/h264 because I need audio - I get audio notifications when there is a notification of new emails/messages coming through from coworkers so I know I need to open the tab to answer. This cannot be done on mjpeg. mjpeg works perfectly but I require the audio which only works on h264 mode. However, I keep getting these black screens and the latency is horrendous, I move the mouse and it does not immediately respond. On mjpeg there is a slight delay for cursor movement but it is reliable and usable. On h264 unusable. Please prioritize this, because not being able to see the screen of the remote device breaks the main use case of PiKVM. mjpeg is not an option due to the lack of audio.

Weird, I never experienced it in an internal network, are you using wired ethernet cable or wifi? The problem is better for me when I'm using it from a different location, so it might depend on location, distance, ISP, etc.

**Comment by plia7** [comment]
> Just wanna say changing the polling rate of the mouse to 10ms instead of the default 100ms greatly improved the pointed accuracy and speed. Consider changing the default to 10ms... and I tried to use webrtc in Firefox and have not had any black screens so far so it seems like you might need to investigate how it's working in Chrome, which is the most popular. For now I am fine using Firefox to get what I need to do my job but I am risking getting told off at work and losing my job because Chrome is the only allowed browser.

Mine is already at 10ms. How using the browser is affecting the host pc?

**Comment by Majestic7979** [comment]
> > Just wanna say changing the polling rate of the mouse to 10ms instead of the default 100ms greatly improved the pointed accuracy and speed. Consider changing the default to 10ms... and I tried to use webrtc in Firefox and have not had any black screens so far so it seems like you might need to investigate how it's working in Chrome, which is the most popular. For now I am fine using Firefox to get what I need to do my job but I am risking getting told off at work and losing my job because Chrome is the only allowed browser.
> 
> Mine is already at 10ms. How using the browser is affecting the host pc?

I don't know why I was having h264 dropouts in Chrome but not in Firefox. I also enabled the USB ethernet with ncm driver in the overrides.yaml and I'm serving wired ethernet via USB to the remote device through the PiKVM and this eliminated my issues. So I think it was a connectivity problem. Probably packet loss which Firefox is somehow handling better than Chrome.

**Comment by plia7** [comment]
> > > Just wanna say changing the polling rate of the mouse to 10ms instead of the default 100ms greatly improved the pointed accuracy and speed. Consider changing the default to 10ms... and I tried to use webrtc in Firefox and have not had any black screens so far so it seems like you might need to investigate how it's working in Chrome, which is the most popular. For now I am fine using Firefox to get what I need to do my job but I am risking getting told off at work and losing my job because Chrome is the only allowed browser.
> 
> > 
> 
> > Mine is already at 10ms. How using the browser is affecting the host pc?
> 
> 
> 
> I don't know why I was having h264 dropouts in Chrome but not in Firefox. I also enabled the USB ethernet with ncm driver in the overrides.yaml and I'm serving wired ethernet via USB to the remote device through the PiKVM and this eliminated my issues. So I think it was a connectivity problem. Probably packet loss which Firefox is somehow handling better than Chrome.

The problem is you can't do keyboard shortcuts in firefox when in full screen unlike chrome.

**Comment by plia7** [comment]
Hi @mdevaev So I actually started to experience the issue again in other location, no internet speed loss in both connecting or host pc... So is it Tailscale again? and what's interesting that in Firefox there are less black screen moments than there are in google chrome... Is it something that can be improved so there is less black screen in chrome like in firefox for when this issue happens? And again, only from an external network and only through h264 this issue happens - I wonder is there is a way to clean the tailscale pipe so the issue stops... As it could work fine for weeks, months, and in one time, somehow have this massive black screen issue? Besides tailscale are there any other supported vpn alternatives?

Thanks.

**Comment by mdevaev** [comment]
I've reported about js error: https://github.com/meetecho/janus-gateway/issues/3283

**Comment by plia7** [comment]
> I've reported about js error: [meetecho/janus-gateway#3283](https://github.com/meetecho/janus-gateway/issues/3283)

Thank you, here is the main "trademark" of the issue, when it blinks it shows this prompt:

<img width="1188" alt="image" src="https://github.com/pikvm/pikvm/assets/134864072/cbaed370-cbbc-4ff6-ab18-01c89f0a0db9">

"H.264 - 1920x1080 / Invalid PeerConnection / 0 fps dynamic"

As you can see the sd memory card icon is still showing in green color (not yellow)

**Comment by mdevaev** [comment]
@plia7 @Majestic7979 I've released small update for this, just a workaround. I don't expect this to significantly improve the situation, but at least it should provide reconnection in case of a bug with unmute. My cursory research shows that this is some kind of bug in Chrome. Since the Janus upstream doesn't know what's going on, I'll try to fix it myself, somehow.

Please update PiKVM OS.

**Comment by Majestic7979** [comment]
Thanks @mdevaev you're a really nice guy and a responsive developer. I installed. I'll test and report back.

**Comment by plia7** [comment]
> @plia7 @Majestic7979 I've released small update for this, just a workaround. I don't expect this to significantly improve the situation, but at least it should provide reconnection in case of a bug with unmute. My cursory research shows that this is some kind of bug in Chrome. Since the Janus upstream doesn't know what's going on, I'll try to fix it myself, somehow.
> 
> 
> 
> Please update PiKVM OS.

Thanks, I'll update at my earliest convenience (although I'm not in the condition to test it at the moment).

**Comment by Majestic7979** [comment]
@mdevaev whatever you did, has cured the issue for me. Thanks again for your hard work on PiKVM. I wanna know if there is any way to improve the latency on h264? When I move the mouse cursor it takes about a second for it to catch up with the grey dot. Is this just how it's always going to be, or are there any optimizations I can do to make the cursor arrow follow the grey dot more responsively? I am also a relatively fast typist and when I am editing configuration files it also takes about a second for a character to appear after I've typed it. Sometimes the Shift key gets stuck and whatever I type is capitalized, and if I hit the dot key . I get a > instead... Shift randomly stays activated (verified through virtual keyboard) when I hold Shift down to capitalize an acronym for example. And sometimes I'll get repeated characters, for example instead of lemon I will get leeeeeeemon.

**Comment by mdevaev** [comment]
The latency when working through a VPN is an unavoidable evil. You can try reducing the mouse polling rate in the system menu and make h264 gop 0. The latency depends on the network conditions, at the moment everything that I could optimize for my part has already been done. I have it at 200ms between continents.

**Comment by plia7** [comment]
> @plia7 @Majestic7979 I've released small update for this, just a workaround. I don't expect this to significantly improve the situation, but at least it should provide reconnection in case of a bug with unmute. My cursory research shows that this is some kind of bug in Chrome. Since the Janus upstream doesn't know what's going on, I'll try to fix it myself, somehow.
> 
> Please update PiKVM OS.

I updated to kvmd-platform-v3-hdmi-rpi4 3.288-1, will try to test when I get a chance and report back. Thanks.

**Comment by Majestic7979** [comment]
@mdevaev unfortunately the problem has returned.

**Comment by mdevaev** [comment]
Well, that only fixed part of it. I will investigate further.

**Comment by Majestic7979** [comment]
Thank you

**Comment by tgrrr** [comment]
Hi,

I have the identical problem. It improved slightly after the update. However, still persists.

> Here is the main "trademark" of the issue, when it blinks it shows this prompt:
> ![image](https://github.com/pikvm/pikvm/assets/4689707/fd88098a-c5b0-434c-86b6-713666fc1eb5)
> "H.264 - 1920x1080 / Invalid PeerConnection / 0 fps dynamic"


Same issue:

> Note: It seems like once a session gets a black screen, it seem to be "infected" with it at that point, it never recovers from it and keeps going back to it so it will keep altering to black screen or slowness even if the screen shows up again and might work for some time.
> As you can see, there is no substantial slowdown in the quality or speed of the internet (speedtest screenshots).

Desktop (please complete the following information):

    OS: [e.g. iOS]
    MacOS 14.1.2
    Browser [e.g. chrome, safari]
    Google Chrome Version 120.0.6099.109 (Official Build) (arm64)
    
    or
    FireFox 120.0.1 (64-bit)
    
    VNC client (if used)
    N/A

Before Update:

PiKVM info:

    Raspberry Pi board version [e.g. RPi 4]
    PiKVM v3 Hat
    PiKVM platform [e.g. v2-hdmi]
    Video capture type [e.g. CSI bridge]
    KVMD version: pacman -Q | grep kvmd
    kvmd 3.287-1
    kvmd-fan 0.26-1
    kvmd-oled 0.26-1
    kvmd-platform-v3-hdmi-rpi4 3.287-1
    kvmd-webterm 0.47-1

    uStreamer version: pacman -Q | grep ustreamer
    ustreamer 5.46-1
    Linux kernel: uname -a
    Linux pikvm 6.1.61-1-rpi-ARCH #1 SMP Fri Nov  3 20:48:52 MSK 2023 armv7l GNU/Linux

After update:

PiKVM info:

    Raspberry Pi board version [e.g. RPi 4]
    PiKVM v3 Hat
    PiKVM platform [e.g. v2-hdmi]
    Video capture type [e.g. CSI bridge]
    KVMD version: pacman -Q | grep kvmd
    kvmd 3.288-1
    kvmd-fan 0.26-1
    kvmd-oled 0.26-1
    kvmd-platform-v3-hdmi-rpi4 3.288-1
    kvmd-webterm 0.47-1
    uStreamer version: pacman -Q | grep ustreamer
    ustreamer 5.46-1
    Linux kernel: uname -a
    Linux pikvm 6.1.61-1-rpi-ARCH #1 SMP Fri Nov  3 20:48:52 MSK 2023 armv7l GNU/Linux    




    Log:

    [2023-12-18 04:48:26 kvmd-janus.service] --- kvmd.apps.janus.runner            INFO --- => == ustreamer/main      -- Creating session 0xb0adc758 ...
[2023-12-18 04:48:26 kvmd-janus.service] --- kvmd.apps.janus.runner            INFO --- => [252830446426366] Creating ICE agent (ICE Full mode, controlling)
[2023-12-18 04:48:27 kvmd-janus.service] --- kvmd.apps.janus.runner            INFO --- => [WARN] [252830446426366] Waiting for candidates-done callback... (slow gathering, are you using STUN or TURN for Janus too, instead of just for users? Consider enabling full-trickle instead)
[2023-12-18 04:48:28 kvmd-janus.service] --- kvmd.apps.janus.runner            INFO --- => [WARN] [252830446426366] Failed to add some remote candidates (added 0, expected 1)
[2023-12-18 04:48:29 kvmd-janus.service] --- kvmd.apps.janus.runner            INFO --- => [252830446426366] The DTLS handshake has been completed
[2023-12-18 04:48:29 kvmd-janus.service] --- kvmd.apps.janus.runner            INFO --- => [WARN] [252830446426366] Failed to add some remote candidates (added 0, expected 1)
[2023-12-18 04:48:29 kvmd.service] --- kvmd.apps.kvmd.streamer           INFO --- => -- INFO  [62753.927    stream] -- H264: Requested keyframe by a sink client
[2023-12-18 04:48:29 kvmd.service] --- kvmd.apps.kvmd.streamer           INFO --- => -- INFO  [62754.019    stream] -- H264: Requested keyframe by a sink client
[2023-12-18 04:48:29 kvmd.service] --- kvmd.apps.kvmd.streamer           INFO --- => -- INFO  [62754.628    stream] -- H264: Requested keyframe by a sink client
[2023-12-18 04:48:30 kvmd.service] --- kvmd.apps.kvmd.streamer           INFO --- => -- INFO  [62754.730    stream] -- H264: Requested keyframe by a sink client
[2023-12-18 04:48:30 kvmd.service] --- kvmd.apps.kvmd.streamer           INFO --- => -- INFO  [62755.297    stream] -- H264: Requested keyframe by a sink client
[2023-12-18 04:48:30 kvmd.service] --- kvmd.apps.kvmd.streamer           INFO --- => -- INFO  [62755.391    stream] -- H264: Requested keyframe by a sink client

PiKVM settings 
![image](https://github.com/pikvm/pikvm/assets/4689707/5ed9d407-0416-4c4e-9ee1-c9b725e68bc8)

**Comment by shamsevan** [comment]
Hi , 
I am experiencing same issue frequently in chrome ! 
while trying to Firefox , the issue seems bit less but it is supper slow there. 
The main frustating part for firefox is , it doesn't accept the keyboard shortcuts , which is making it pretty difficult to run and execute tasks. please give me a solution

**Comment by blackandcold** [comment]
Have the same problem - strange it happens in Vivaldi (Chrome) and NOT in Firefox

**Comment by mdevaev** [comment]
Is anyone who has this problem playing often and constantly ready to provide me with a TeamViewer session to their computer so that I can look at it?

**Comment by plia7** [comment]
> Is anyone who has this problem playing constantly ready to provide me with a TeamViewer session to their computer so that I can look at it?

Hi @mdevaev, 
@shamsevan says the issue happens 10 times in an hour and actually after every few minutes, he's on discord and says he could let you remote to his machine to investigate (he uses the latest os version with the partial fix).

Thank you.

**Comment by Majestic7979** [comment]
@mdevaev I found an interesting thing that might help you. I'm using the latest build of PiKVM as of writing this (ran your curl command) so you know which version. Basically, until there is some audio playback the screen will keep cutting out. But the moment there's some audio playback, it stops and stays stable. I can reproduce this every single time. How I do is I logon to Windows while it's in JPEG mode, then I switch to WebRTC and it starts cutting out. I open the Windows volume slider (which makes a sound when you move and release it) and as soon as I hear the Windows make a sound, I get a stable connection. So it seems like until it's silent the screen cuts out!! I use WebRTC for the audio capability over HDMI too, which JPEG mode does not support. Please note that when I swap to WebRTC I have to open the PiKVM menu and move the volume to 100%. I am talking about the Windows volume slider just to make some sound. As long as the remote machine sends a sound the issue will stop. So that narrows it down further for you to investigate.

**Comment by mdevaev** [comment]
Do I understand correctly that even if the sound is turned on in PiKVM, this is not enough, and you need the host to start playing the sound itself?

**Comment by Majestic7979** [comment]
> Do I understand correctly that even if the sound is turned on in PiKVM, this is not enough, and you need the host to start playing the sound itself?

That's what I was thinking until today but for some reason it was glitching again. I waited a few days just to make sure before making that comment, but today I had issues again. It does seem related to the audio. I'll be monitoring some more to see if it was just "luck" or if there really is a link to audio causing issues.

**Comment by RSATom** [comment]
@mdevaev Can you please clarify, setup you are using to access streams from Janus server doesn't use TURN server at all, right?

**Comment by mdevaev** [comment]
@RSATom yes, only STUN.

**Comment by RSATom** [comment]
Ok, did you see something like "ICE state changed to failed" in JS console right before issue?

**Comment by mdevaev** [comment]
@RSATom this is a first option. The second is many "remote track muted", see above.

**Comment by RSATom** [comment]
I think there are multiple issues here. So it's required to split them and only after try to fix.

So, first of all it's required to understand how to simplify reproducing this issue. And I think it should be simplest part - if there is firewall present - one can just disable UDP packets passing through firewall and wait ICE disconnected (~1 second), ICE failed (~2 seconds).

Then browser client will try to restore connection. And there are 2 ways to restore WebRTC connection: full reconnect and ICE renegotiation. Since you are using custom plugin I can't say if it supports the second one (please clarify).

Other thing it's required to take into account - full reconnect with Janus can be tricky, especially if you don't detach from Janus plugin on Peer Connection drop.

Maybe all above already known to you, please tell me. I just need to understand what exactly you already tried.

**Comment by mdevaev** [comment]
> Since you are using custom plugin I can't say if it supports the second one (please clarify).

No, it's not supported: https://github.com/pikvm/ustreamer/blob/master/janus/src/plugin.c

Although I wrote this, I have a fairly superficial idea of the internal structure of Janus. If there are any problems, I completely terminate the connection with Janus and re-establish it:

https://github.com/pikvm/kvmd/blob/master/web/share/js/kvm/stream_janus.js

Perhaps this is really due to packet loss, but users say that this problem either does not reproduce in Firefox or occurs much less frequently than in Chrome.

If it's convenient for you, let's get in touch via PM at discord: https://discord.gg/bpmXfz5 (mdevaev there)

**Comment by RandomJerk** [comment]
I can confirm that Firefox works better and does not see black screen interruptions.  The issue happens on Chrome and Chromium based browsers

**Comment by plia7** [comment]
> I can confirm that Firefox works better and does not see black screen interruptions.  The issue happens on Chrome and Chromium based browsers

Yes, maybe Firefox works a little better, but it doesn't support keyboard shortcuts , which is making it pretty difficult to run and execute tasks so it's not really a suitable workaround.

**Comment by RSATom** [comment]
did anybody look into `chrome://webrtc-internals`?

**Comment by plia7** [comment]
> did anybody look into `chrome://webrtc-internals`?

I believe @mdevaev looked in mine few months ago when I had the issue. Or possibly he looked again recently in @shamsevan machine.

**Comment by RSATom** [comment]
I just got an Idea it can be broken by some reason h264 video stream (i.e. ICE connection is fine)... And working Firefox maybe confirms it... Firefox and Chrome have different codecs implementation inside.

**Comment by RSATom** [comment]
@plia7 everything is possible. But to say more it's required to know if there is `failed` state happens for ICE before black screen. I'm still waiting my HDMI-to-CSI device to start debugging issue myself. I hope it will arrive on this weekend...

**Comment by RSATom** [comment]
Firefox is slowly dying, so yes, it's not a solution in any case...

**Comment by Majestic7979** [comment]
Based on comment above I turned off H/w video decoding in chrome flags, and it makes no difference. I also noticed that in Firefox my mouse movements are more fluid, more responsive. The delay appears to be much less, so if I move the mouse from one corner of the screen to the opposite corner, I can see my mouse moving faster in Firefox, in Chrome there is a terrible delay. So Chrome is unusable in its own way, but it's the browser I need to use sometimes due to customer restrictions in their IT.

**Comment by plia7** [comment]
> Based on comment above I turned off H/w video decoding in chrome flags, and it makes no difference. I also noticed that in Firefox my mouse movements are more fluid, more responsive. The delay appears to be much less, so if I move the mouse from one corner of the screen to the opposite corner, I can see my mouse moving faster in Firefox, in Chrome there is a terrible delay. So Chrome is unusable in its own way, but it's the browser I need to use sometimes due to customer restrictions in their IT.

Did you also try to toggle off "use graphics acceleration when available" setting?

I wonder why the mouse movement and usability appears to be slower in firefox, I guess that's a separate issue.

Thanks.

**Comment by plia7** [comment]
> @plia7 everything is possible. But to say more it's required to know if there is `failed` state happens for ICE before black screen appear. I'm still waiting my HDMI-to-CSI device to start debugging issue myself. I hope it will arrive on this weekend...

Looking at the logs on this ticket, I don't see ICE failed state, but it's maybe happening on the first time which causes it to go into this black screen loop (and logs typically gathered after that), so yes, maybe it's good to have full log for it, to make sure 100% if it happens or not so then a proper ticket can be submitted on the chromium bug report website (if it actually turns out not to be pikvm's fault in this case).

Thank you.

**Comment by RSATom** [comment]
Finally I've got my HDMI-2-CSI device, did some debugging, and found with current implementation `failed` ICE state never happens. I think the reason is related to handling `disconnected` ICE state by browser and `janus.js`:
1. When peer connection is established and running, UDP packets can stop coming by some reason, and after some amount of missed packets, ICE State switches to [disconnected](https://developer.mozilla.org/en-US/docs/Web/API/RTCIceTransport/state#disconnected) state.
_Important note about `disconnected` ICE state: the name of this state can be confusing but it's not final state. If packets will start flow again - state will switch to `connected` state again. And only if there will be no packets too long amount time state will be switched to `failed` state. And it's final state._
2. Near the same time WebRTC library sends `mute` event to track. And `janus.js` handles it by scheduling [track remove](https://github.com/meetecho/janus-gateway/blob/63ee713102f453447e3c912a7bb45791d4d198c5/html/janus.js#L1926-L1940) from **internal Janus.js structures** (but it doesn't mean track disappears from RTCPeerConnection, it's only `janus.js` internal logic)
3. If there is no `unmute`, after timeout `janus.js` removes track and [sends](https://github.com/meetecho/janus-gateway/blob/63ee713102f453447e3c912a7bb45791d4d198c5/html/janus.js#L1933) `remotestream` event. And this is the point where `stream_janus.js` gets confusing [onremotestream](https://github.com/pikvm/kvmd/blob/4690e330881fcd2ccb30a2ce154be0fdbf1990db/web/share/js/kvm/stream_janus.js#L222) without video tracks. I.e. it's intended behavior, and there is no reason to disconnect from janus at this point. The right time to start reconnect logic is `failed` ICE state (or even better `failed` state of [RTCPeerConnection.connectionstate](https://developer.mozilla.org/en-US/docs/Web/API/RTCPeerConnection/connectionState#failed)

Unfortunately all above has nothing to do with "black screen" issue. When I enabled UDP packets flow on my test environment - reconnect to Janus happened successfully and I was able to see incoming video...

**Comment by plia7** [comment]
> Finally I've got my HDMI-2-CSI device, did some debugging, and found with current implementation `failed` ICE state never happens. I think the reason is related to handling `disconnected` ICE state by browser and `janus.js`:
> 
> 1. When peer connection is established and running, UDP packets can stop coming by some reason, and after some amount of missed packets, ICE State switches to [disconnected](https://developer.mozilla.org/en-US/docs/Web/API/RTCIceTransport/state#disconnected) state.
>    _Important note about `disconnected` ICE state: the name of this state can be confusing but it's not final state. If packets will start flow again - state will switch to `connected` state again. And only if there will be no packets for too long amount time state will be switched to `failed` state. And it's final state._
> 2. Near the same time WebRTC library sends `mute` event to track. And `janus.js` handles it by scheduling [track remove](https://github.com/meetecho/janus-gateway/blob/63ee713102f453447e3c912a7bb45791d4d198c5/html/janus.js#L1926-L1940) from **internal Janus.js structures** (but it doesn't mean track disappears from RTCPeerConnection, it's only `janus.js` internal logic)
> 3. If there is no `unmute`, after timeout `janus.js` removes track and [sends](https://github.com/meetecho/janus-gateway/blob/63ee713102f453447e3c912a7bb45791d4d198c5/html/janus.js#L1933) `remotestream` event. And this is the point where `stream_janus.js` gets confusing [onremotestream](https://github.com/pikvm/kvmd/blob/4690e330881fcd2ccb30a2ce154be0fdbf1990db/web/share/js/kvm/stream_janus.js#L222) without video tracks. I.e. it's intended behavior, and there is no reason to disconnect from janus at this point. The right time to start reconnect logic is `failed` ICE state (or even better `failed` state of [RTCPeerConnection.connectionstate](https://developer.mozilla.org/en-US/docs/Web/API/RTCPeerConnection/connectionState#failed) (but obviously it can keep black screen for longer period of time)
> 
> Unfortunately all above has nothing to do with "black screen" issue. When I enabled UDP packets flow on my test environment - reconnect to Janus happened successfully and I was able to see incoming video...

Ok so what this all mean? Do you need the conditions to reproduce the issue to analyze it? Is it possible for you to connect with one of the pikvm users who reported the issue/have the issue and can reproduce it such as through a TeamViewer session to try to analyze the issue on their end (which is what @mdevaev tried to do but couldn't determine the root cause)? Or what is it that you need that will help you to get to the root cause of the issue? Do we need to send HDMI-2-CSI device to one of the users who have the issue? 

Keep in mind it's a very specific set of conditions to reproduce the issue that includes being on an external network and connecting to tailscale , so I'm not sure if you are testing all of this locally with a physical pikvm without actually connecting to a remote pikvm through tailscale.
So either you need to connect to a remote pikvm (on an external network) through tailscale and see if you could get the intermittent black screen at some point and then start analyzing it with the HDMI-2-CSI device or maybe some of the other members in this post have more suggestions or could help with providing you with the conditions needed to reproduce the issue.

Thanks.

**Comment by RSATom** [comment]
@plia7 I just shared my findings, nothing more at this moment. HDMI-2-CSI - is device all pikvm users already have actually. It's where you put you HDMI cable :)
Also, I've found another one possible issue and working with @mdevaev on it.

**Comment by plia7** [comment]
@mdevaev reached out to me and asked if someone currently using the external network with tailscale and is able to reproduce the issue - @RandomJerk @blackandcold @tgrrr or anyone else who read this - Are you available to assist through a TeamViewer session so the issue can be inspected/further analyzed by @mdevaev and @RSATom? 

Thank you.

**Comment by RSATom** [comment]
it would be great if someone could provide me remote access to pikvm with reproducible issue. I don't need real server, I just need video stream. It can be even some tv stick attached to pikvm/

**Comment by plia7** [comment]
> **Additional context**
> This is the same system setup I used to have in my local network with the connecting pc and host pc - and I never encountered this issue when everything was in the loca

I found 10 people who have the issue, we created a group in Discord. Please come and join us.

**Comment by Majestic7979** [comment]
I feel I need to just remind all reading at this stage that I do not use tailscale. I am accessing the PiKVM from within the internal network... so I'm on 192.168.1.20 and the PiKVM is 192.168.1.21, for example. I get the issue without any VPN or tunneling.

**Comment by Majestic7979** [comment]
> it would be great if someone could provide me remote access to pikvm with reproducible issue. I don't need real server, I just need some video stream. It can be even some tv stick attached to pikvm...

Are you able to send me your email address? I will get in touch.

**Comment by RSATom** [comment]
@Majestic7979 rsatom_at_gmail_dot_com

**Comment by RSATom** [comment]
> I feel I need to just remind all reading at this stage that I do not use tailscale. I am accessing the PiKVM from within the internal network... so I'm on 192.168.1.20 and the PiKVM is 192.168.1.21, for example. I get the issue without any VPN or tunneling.

Are 192.168.1.20 and 192.168.1.21 connected to the same Ethernet switch?

**Comment by RSATom** [comment]
> Thanks, you can remove your email now, I've noted it. I'll be writing soon.

:+1: but there is no need to do it. It published on my github profile anyway :smiley:

**Comment by Majestic7979** [comment]
> > Thanks, you can remove your email now, I've noted it. I'll be writing soon.
> 
> 👍 but there is no need to do it. It shown on my github profile anyway 😃
> 
> > I'll be bringing the PiKVM to my house, where I have it connected to my UniFi network. It will be connected to a UniFi Flex Mini, that will connect to a USW-24-PoE, and that in turn connected to a UniFi Dream Machine. But it's all the same network within the router configuration.
> 
> and there is no any WiFi in between, right?

I have an Access Point of course, only for the wireless devices. Not for the PiKVM and not for the PC I'll be using because they can be wired and that's how I prefer it.

**Comment by Majestic7979** [comment]
I emailed you now, @RSATom

**Comment by Majestic7979** [comment]
As a means of updating everyone I'm writing here. @RSATom logged in to my PC via zoom yesterday and showed me how to capture the data he needed to analyze the issue. He now has all the necessary logs, so hopefully some sort of fix can come out of it. I want to thank him for helping and being patient.

**Comment by Majestic7979** [comment]
> @Majestic7979 unfortunately I don't have good news for you. According to logs your browser just didn't get UDP packets periodically. So it's network (more likely) or pikvm side (less likely) issue. Did you use VPN yesterday when did issue reproducing?

@RSATom No VPN. Oh :( But it's weird that it doesn't happen on Firefox, only on Chrome... so while it may be the case that network can play a role in this issue, the way it's handled is different. Isn't there a way to code that janus server to stop cutting out? What if it just waits for more UDP packets? I don't know how it works so of course it's just me speculating...

**Comment by RSATom** [comment]
> But it's weird that it doesn't happen on Firefox, only on Chrome... 
Yes, it's confusing me too...

And UDP packets are sending always while peer connection is active, without waiting something...

**Comment by Majestic7979** [comment]
> > But it's weird that it doesn't happen on Firefox, only on Chrome...
> 
> Yes, it's confusing me too...
> 
> And UDP packets are sending always while peer connection is active, without waiting something...

Right, I thought so, just like a IP CCTV camera works. But in this case the connection is being cut out. I can see when it goes black on the console that it says the peer connection is down. So on chrome something is happening to break the connection... Maybe this is happening in Firefox too, however it doesn't cause any black screens. On Firefox occasionally I get the hiccup on the video where it goes slower but it never completely goes black and cuts out.

**Comment by RSATom** [comment]
can you please try run `ping -t pikvm_ip` (if browser is running on windows)  from host where browser is running to pikvm (or router in front of it)?

also, if Firefox experiencing the same issue you should be able to see `slowlink` event in JS console.

**Comment by RSATom** [comment]
> like a IP CCTV camera works

if it uses RTSP protocol - yes, very similar, but RTSP can send data over TCP in some cases.

**Comment by Majestic7979** [comment]
> can you please try run `ping -t pikvm_ip` (if browser is running on windows) from host where browser is running to pikvm (or router in front of it)?
> 
> also, if Firefox experiencing the same issue you should be able to see `slowlink` event in JS console.

```
ping -t 10.0.0.225

Pinging 10.0.0.225 with 32 bytes of data:
Reply from 10.0.0.225: bytes=32 time<1ms TTL=64
Reply from 10.0.0.225: bytes=32 time<1ms TTL=64
Reply from 10.0.0.225: bytes=32 time<1ms TTL=64
Reply from 10.0.0.225: bytes=32 time=1ms TTL=64
Reply from 10.0.0.225: bytes=32 time<1ms TTL=64
Reply from 10.0.0.225: bytes=32 time=1ms TTL=64
Reply from 10.0.0.225: bytes=32 time<1ms TTL=64
Reply from 10.0.0.225: bytes=32 time<1ms TTL=64
Reply from 10.0.0.225: bytes=32 time<1ms TTL=64
Reply from 10.0.0.225: bytes=32 time<1ms TTL=64
Reply from 10.0.0.225: bytes=32 time<1ms TTL=64
Reply from 10.0.0.225: bytes=32 time<1ms TTL=64
Reply from 10.0.0.225: bytes=32 time=14ms TTL=64
Reply from 10.0.0.225: bytes=32 time<1ms TTL=64
Reply from 10.0.0.225: bytes=32 time<1ms TTL=64
Reply from 10.0.0.225: bytes=32 time<1ms TTL=64
Reply from 10.0.0.225: bytes=32 time<1ms TTL=64
Reply from 10.0.0.225: bytes=32 time<1ms TTL=64

Ping statistics for 10.0.0.225:
    Packets: Sent = 18, Received = 18, Lost = 0 (0% loss),
Approximate round trip times in milli-seconds:
    Minimum = 0ms, Maximum = 14ms, Average = 0ms
Control-C
```

I am filtering for "slowlink" in the Firefox console. I'll come back to you soon saying if there was something or not.

**Comment by RSATom** [comment]
it's better keep running ping for long period of time and check output it when you'll get black screen issue.

**Comment by Majestic7979** [comment]
> > like a IP CCTV camera works
> 
> if it uses RTSP protocol - yes, very similar, but RTSP can send data over TCP in some cases.

I understand that PiKVM wants UDP for the lower latency. Is it possible to implement also TCP option (user can configure in override.yaml) for the janus server to see if that could resolve? At a price of higher latency but more reliability... It's a mystery isn't it?

**Comment by Majestic7979** [comment]
> it's better keep running ping for long period of time and check output it when you'll get black screen issue.

Ah I see. No problem, I will keep it running.

**Comment by Majestic7979** [comment]
Got the slowlink on Firefox now:
```
Object { janus: "slowlink", session_id: 7663056455694392, sender: 86363667049486, media: "video", uplink: true, lost: 10 }
​
janus: "slowlink"
​
lost: 10
​
media: "video"
​
sender: 86363667049486
​
session_id: 7663056455694392
​
uplink: true
​
<prototype>: Object { … }
janus.js:749:10
```
And the ping:
```
ping -t 10.0.0.225

Pinging 10.0.0.225 with 32 bytes of data:
Reply from 10.0.0.225: bytes=32 time<1ms TTL=64
Reply from 10.0.0.225: bytes=32 time<1ms TTL=64
Reply from 10.0.0.225: bytes=32 time<1ms TTL=64
Reply from 10.0.0.225: bytes=32 time<1ms TTL=64
Reply from 10.0.0.225: bytes=32 time<1ms TTL=64
Reply from 10.0.0.225: bytes=32 time=1ms TTL=64
Reply from 10.0.0.225: bytes=32 time<1ms TTL=64
Reply from 10.0.0.225: bytes=32 time<1ms TTL=64
Reply from 10.0.0.225: bytes=32 time=9ms TTL=64
Reply from 10.0.0.225: bytes=32 time<1ms TTL=64
Reply from 10.0.0.225: bytes=32 time<1ms TTL=64
Reply from 10.0.0.225: bytes=32 time<1ms TTL=64
Reply from 10.0.0.225: bytes=32 time<1ms TTL=64
Reply from 10.0.0.225: bytes=32 time<1ms TTL=64
Reply from 10.0.0.225: bytes=32 time<1ms TTL=64
Reply from 10.0.0.225: bytes=32 time=1ms TTL=64
Reply from 10.0.0.225: bytes=32 time=2ms TTL=64
Reply from 10.0.0.225: bytes=32 time=5ms TTL=64
Reply from 10.0.0.225: bytes=32 time=5ms TTL=64
Reply from 10.0.0.225: bytes=32 time=4ms TTL=64
Reply from 10.0.0.225: bytes=32 time=5ms TTL=64
Reply from 10.0.0.225: bytes=32 time<1ms TTL=64
Reply from 10.0.0.225: bytes=32 time<1ms TTL=64
Reply from 10.0.0.225: bytes=32 time<1ms TTL=64
Reply from 10.0.0.225: bytes=32 time<1ms TTL=64
Reply from 10.0.0.225: bytes=32 time=5ms TTL=64
Reply from 10.0.0.225: bytes=32 time=5ms TTL=64
Reply from 10.0.0.225: bytes=32 time=5ms TTL=64
Reply from 10.0.0.225: bytes=32 time=5ms TTL=64
Reply from 10.0.0.225: bytes=32 time=4ms TTL=64
Reply from 10.0.0.225: bytes=32 time=5ms TTL=64
Reply from 10.0.0.225: bytes=32 time=1ms TTL=64
Reply from 10.0.0.225: bytes=32 time<1ms TTL=64
Reply from 10.0.0.225: bytes=32 time<1ms TTL=64
Reply from 10.0.0.225: bytes=32 time<1ms TTL=64
Reply from 10.0.0.225: bytes=32 time<1ms TTL=64
Reply from 10.0.0.225: bytes=32 time<1ms TTL=64
Reply from 10.0.0.225: bytes=32 time=1ms TTL=64
Reply from 10.0.0.225: bytes=32 time<1ms TTL=64
Reply from 10.0.0.225: bytes=32 time<1ms TTL=64
Reply from 10.0.0.225: bytes=32 time<1ms TTL=64
Reply from 10.0.0.225: bytes=32 time<1ms TTL=64
Reply from 10.0.0.225: bytes=32 time<1ms TTL=64
Reply from 10.0.0.225: bytes=32 time<1ms TTL=64
Reply from 10.0.0.225: bytes=32 time<1ms TTL=64
Reply from 10.0.0.225: bytes=32 time<1ms TTL=64

Ping statistics for 10.0.0.225:
    Packets: Sent = 46, Received = 46, Lost = 0 (0% loss),
Approximate round trip times in milli-seconds:
    Minimum = 0ms, Maximum = 9ms, Average = 1ms
Control-C
```
I will continue to run the ping if you need much more info than that. The black screen happens the moment I start using Chrome... it's almost always instant issue.

**Comment by RSATom** [comment]
@mdevaev if I understand correctly jpeg mode uses TCP, right?

@Majestic7979 if I remember correctly Janus doesn't have TCP support. But it's still possible to get it with TURN server.

**Comment by Majestic7979** [comment]
> @mdevaev if I understand correctly jpeg mode uses TCP, right?
> 
> @Majestic7979 if I remember correctly Janus doesn't have TCP support. But it's still possible to get it with TURN server.

I'll have to leave the technical details to you cleverer boys :)

**Comment by RSATom** [comment]
hm... there is no packets loss with ping...

**Comment by Majestic7979** [comment]
I hope I don't sound like an asshole, it's not my intention but if there is a technical way to try the TCP could we do that? Just maybe to rule out that solution? I can see that you asked Maxim, hopefully he can let us know 🤞🏻

**Comment by RSATom** [comment]
Did you see something like picture freezes in Firefox? it's possible it just don't switch to black screen on `video` tag detach from WebRTC stream. But i'm not sure...

**Comment by Majestic7979** [comment]
> Did you see something like picture freezes in Firefox? it's possible it just don't switch to black screen on `video` tag detach from WebRTC stream. But i'm not sure...

On Firefox I sometimes see slowness in the picture, for example when I move the mouse cursor it will have a delay between the cursor moving and the final position of the blue dot. But never ever black screen.

This is how I have it set up on Firefox:
![image](https://github.com/pikvm/pikvm/assets/128995501/4800d310-609a-47e6-9392-1c910c814f00)

**Comment by Majestic7979** [comment]
One thing that I noticed is that on Firefox I have only received 2 slowlinks so far... On Chrome the issue happens pretty much every few seconds...

**Comment by RSATom** [comment]
> I hope I don't sound like an asshole, it's not my intention but if there is a technical way to try the TCP could we do that? 

Don't worry, everything is ok. 

I see Janus supports [TURN server directly](https://github.com/meetecho/janus-gateway/blob/a0175e89735bacfee4e5bc2185ec2e05f02a7358/conf/janus.jcfg.sample.in#L340)

So maybe simplest way to try it is find/install/buy TURN server and add it to config.

**Comment by RSATom** [comment]
> One thing that I noticed is that on Firefox I have only received 2 slowlinks so far... On Chrome the issue happens pretty much every few seconds...

Hm... It's very strange and I can't explain it right now....

**Comment by Majestic7979** [comment]
If you can send me instructions I'll follow, to ensure I'm not doing something I think is correct but is totally wrong 😂

**Comment by Majestic7979** [comment]
Also:
```
Ping statistics for 10.0.0.225:
    Packets: Sent = 1919, Received = 1919, Lost = 0 (0% loss),
Approximate round trip times in milli-seconds:
    Minimum = 0ms, Maximum = 16ms, Average = 0ms
Control-C
```

**Comment by RSATom** [comment]
@Majestic7979 you can try https://xirsys.com/

**Comment by Majestic7979** [comment]
> @Majestic7979 you can try https://xirsys.com/
> 
> If I remember correctly they provided free accounts some way.
> 
> And it's ok to contact me if you'll need assistance to configure your pikvm...

I understand you're sleeping at this time so I'll check back later.

I installed "CoTURN" in docker making sure the relevant ports are exposed, I am ready to test and see what happens. How to proceed to add the TURN server I'm running locally? I understand it may not be functional but I want to know how to set up the PiKVM at least. I'll test and see if I can make it work. If not, I appreciate the open invite to get your assistance :)

**Comment by Majestic7979** [comment]
Another question - if both devices are in the same network/subnet, why do I need a TURN server at all!? 

I mentioned in the past that I am using this:
https://docs.pikvm.org/usb_ethernet/

So the device connected to the PiKVM is receiving networking from the PiKVM via USB.

The network subnet will therefore be different. Is this what may be causing the problem? I have used USB networking because it eliminates one more cable and the bandwidth is acceptable over USB since the server in question is for monitoring purposes, it does not do any heavy data transfers.

**Comment by Majestic7979** [comment]
Just some food for thought... Why is Janus the preferred solution instead of something like go2rtc? Could testing with a different webrtc server maybe offer a solution? Only trying to look at a way of solving this that may be obvious but not yet thought about...

**Comment by mdevaev** [comment]
Sswitching to another rtc gateway will lead to new problems that did not exist with janus.

**Comment by RSATom** [comment]
@Majestic7979 you can find Janus config on pikvm at `/etc/kvmd/janus/janus.jcfg`

to configure TURN you should add related lines [under the `nat` section](https://github.com/meetecho/janus-gateway/blob/a0175e89735bacfee4e5bc2185ec2e05f02a7358/conf/janus.jcfg.sample.in#L331-L344) Unfortunately I never used Janus with NAT configured that way myself (usually it used on client side), so can't say how/if it will work.

> Why is Janus the preferred solution instead of something like go2rtc? Could testing with a different webrtc server maybe offer a solution?

1. It will take some time to implement support of some other WebRTC library
2. There is no guarantee it will be better
3. It's required find exact reason of problem or it will be just waste of time.

**Comment by Majestic7979** [comment]
I'm gonna sit and wait for you guys to explore solutions for these reasons:

1. you are programmers, I am not, so your technical knowledge is vastly beyond mine
2. I may not have configured the TURN server correctly, it is likely that it is incorrect as I run services behind Traefik reverse proxy
3. I don't want to introduce another layer of issues for us to worry about
4. PiKVM is a turnkey solution so the developer knows all the code and how to deal with it, the developer is the best people to deal with it

Thanks guys, I'll just be patient and let you work on this. I hope you can find a solution for all of us your fans and users sooner rather than later... Wishing you all good luck! I will be watching in case I can provide further info but otherwise just taking myself out of the game, let me know if you need anything from me :-D

**Comment by Majestic7979** [comment]
Some extra info for you @RSATom, you are aware that I was using the Ethernet over USB, so the subnet was not the same as the main network where I'm accessing from the browser. When I connected the device to WiFi (to test on the same subnet) the black screens in Chrome completely stopped. So yes, this appears to be a NAT issue. But I run something called Frigate NVR in my home lab, this is for security cameras. Frigate uses an implementation go2rtc for WebRTC streaming. If I access from my 5G phone completely bypassing my subnet, it works flawless for streaming UDP cameras (I mean flawless in the sense that the cameras connect normally and stream normally, but PiKVM is not a camera, so it could be different).

I'd be interested to test your implementation of Gstreamer to see if this issue goes away. I really wanted to try your potential solution of TURN server but I don't think I can configure CoTURN in my home lab behind my reverse proxy in a correct manner so I don't want to introduce another issue in the mix... Trying to keep it as simple as possible to help you :)

**Comment by Majestic7979** [comment]
> @Majestic7979 I'm a little bit confused with your network configuration... So let me ask you some questions:
> 
> 1. You had USB to Ethernet adapter connected directly to your pikvm, right?
> 2. Why you had to use USB to Ethernet adapter with pikvm? isn't it have Ethernet on board?

Ok I'll clarify, I'll try to provide the detail how it makes sense.

I have Device I want to manage, this device has HDMI output plugged into input of PiKVM so I can see the device screen. I have USB-C connected from the device to the PiKVM so I can do mouse and keyboard from PiKVM in the browser. I have only 1 ethernet cable plugged into the PiKVM. I think I confused you. I did not use a USB adaptor to provide ethernet to the PiKVM, I plugged the ethernet cable directly into the PiKVM and I am using instructions from here to pass through the ethernet to the device via the PiKVM (because I only have 1 ethernet cable in this location): https://docs.pikvm.org/usb_ethernet/

This device was originally connected via WiFi for simplicity, because wiring an extra cable to the device was going to have higher costs. I could not fit an extra switch. I have a VPN into that device so I can monitor it remotely via SSH. Because the WiFi is meshed (Ubiquiti UniFi) any brief roaming events from one AP to another due to signal variations would cause the VPN to disconnect and break the tunnel. So I had to drop WiFi. A solution I found was to use the Ethernet over USB on the PiKVM, so I could use the single ethernet cable to have the PiKVM on the wired ethernet, and then share the connection over USB (https://docs.pikvm.org/usb_ethernet/).

If you follow all the steps on the above instructions, you can successfully provide ethernet via USB with internet access, but the PiKVM will be on Subnet A and the connection provided to the device will be on Subnet B (because the instructions themselves say it can't be on the same subnet of the PiKVM).

Therefore as it's not on the same subnet, it's doing NAT (as I understand). But all devices are managed in the same network (as in the same router). Maybe a way to think of it is to pretend that the PiKVM is on VLAN 1, the device is on VLAN 2.

Does this all make sense to you so far?

So what have I done: I disabled the OTG Ethernet over USB on the Override.yaml inside /etc/kvmd and I reconnected the device to WiFi like it was before. And the black screens are totally gone. Because now the PiKVM and the device are on the same subnet (or VLAN 1). On the same subnet the issue does not occur. On different subnets/networks the black screen occurs. The problem I have now is that if I use WiFi, I get VPN disconnections that break the tunnel. If I use USB over Ethernet through the PiKVM to share with the device, I get black screens. The physical solution is to pull an extra cable (I can't add a switch). But that still does not fix the issue with the PiKVM not handling this type of situation, which now make sense when people are saying it happens on Tailscale because it's doing NAT as well. So basically, the workaround of pulling a new cable won't fix the issue for everyone else.

**Comment by RSATom** [comment]
@Majestic7979 Now I think I understood... did you try to touch your USB to Ethernet device after some period of time? is it hot?

**Comment by Majestic7979** [comment]
> @Majestic7979 Now I think I understood... did you try to touch your USB to Ethernet device after some period of time? is it hot?

There's no USB to ethernet device :-P The PiKVM has an ethernet cable coming directly from the router plugged into the ethernet port. If you open the pikvm document I linked (https://docs.pikvm.org/usb_ethernet/) you can see there is a way to make the PiKVM provide a virtual ethernet to the device, just like it provides a virtual keyboard and mouse... So the PiKVM is is just passing through the ethernet (sharing it) with the device as a virtual ethernet via USB the same way the keyboard and mouse is provided to the device virtually via USB... the device gets a virtual ethernet connection... There is no physical USB to ethernet device anywhere in this process, the PiKVM is providing keyboard, mouse and ethernet devices virtually over the USB connection...

**Comment by RSATom** [comment]
Ah... sorry, got it now.

**Comment by Majestic7979** [comment]
If you have a PiKVM using ethernet please follow the instructions and you will see that the device connected to the PiKVM will get a new ethernet interface, but it's coming via USB from the PiKVM :-)

**Comment by Majestic7979** [comment]
But that virtual ethernet must be (according to the instructions) on a different subnet...

**Comment by RSATom** [comment]
Got it. Thanks! I need some time to understand how it works internally to understand why it breaks pikvm....

**Comment by Majestic7979** [comment]
Just to clarify the above, you can see that the virtual ethernet is assigning to the device an IP in subnet 10.0.1.x whereas the browser is running on a computer on the subnet 10.0.0.x

**Comment by RSATom** [comment]
> whereas the browser is running on a computer on the subnet 10.0.0.x

i.e. computer with browser is running on the same subnet where pikvm connected physically with ethernet cable, right?

**Comment by RSATom** [comment]
@plia7 are you using https://docs.pikvm.org/usb_ethernet/ on your pikvm?

**Comment by Majestic7979** [comment]
> > whereas the browser is running on a computer on the subnet 10.0.0.x
> 
> @Majestic7979 i.e. computer with browser is running on the same subnet where pikvm connected physically with ethernet cable, right?

Yes correct. And to clarify if I am using the passthrough ethernet as explained above then the remote device gets a different subnet from the PiKVM virtual ethernet adaptor. But the PiKVM itself is on the same subnet as the device used to run Chrome.

**Comment by plia7** [comment]
> > > @plia7 are you using https://docs.pikvm.org/usb_ethernet/ on your pikvm?
> 
> > 
> 
> > @RSATom No, I'm using the physical ethernet port and it works fine. Should we move @Majestic7979 thread/comments to a new ticket given it's only caused by his usb virtual ethernet (and it works fine for him otherwise)? Since it's not really related at this point to the original issue?
> 
> > 
> 
> > Suggestion: Use his existing ticket #1170 and rename the subject to "janus.js on Google Chrome black screen, or streaming cuts off periodically **when using usb virtual ethernet in local network**"
> 
> > 
> 
> > @RSATom The issue seem like the same symptoms but looks like two different root causes given the setup difference to get these symptoms. Hopefully resolving one of them should shed some light on the other one.
> 
> > 
> 
> > Thanks.
> 
> 
> 
> I am not against your idea, I hope it does not mean my issue gets less attention... Because this does seem related to a network route problem/NAT problem in both instances... whatever is decided I just ask that my issue is not forgotten, thank you 🙏🏻

Let's let @RSATom decide if it's related or not, in both cases I suppose it should work, should it my ticket gets more attention and it leads at the end to some enlightenment with regards to the original issue I don't mind for you to keep using this thread 😀

**Comment by Majestic7979** [comment]
> > > > @plia7 are you using https://docs.pikvm.org/usb_ethernet/ on your pikvm?
> > 
> > 
> > > 
> > 
> > 
> > > @RSATom No, I'm using the physical ethernet port and it works fine. Should we move @Majestic7979 thread/comments to a new ticket given it's only caused by his usb virtual ethernet (and it works fine for him otherwise)? Since it's not really related at this point to the original issue?
> > 
> > 
> > > 
> > 
> > 
> > > Suggestion: Use his existing ticket #1170 and rename the subject to "janus.js on Google Chrome black screen, or streaming cuts off periodically **when using usb virtual ethernet in local network**"
> > 
> > 
> > > 
> > 
> > 
> > > @RSATom The issue seem like the same symptoms but looks like two different root causes given the setup difference to get these symptoms. Hopefully resolving one of them should shed some light on the other one.
> > 
> > 
> > > 
> > 
> > 
> > > Thanks.
> > 
> > 
> > I am not against your idea, I hope it does not mean my issue gets less attention... Because this does seem related to a network route problem/NAT problem in both instances... whatever is decided I just ask that my issue is not forgotten, thank you 🙏🏻
> 
> Let's let @RSATom decide if it's related or not, in both cases I suppose it should work, so if my ticket gets more attention and helps to resolve yours and it leads at the end to some enlightenment with regards to the original issue I don't mind for you to keep using this thread as we both benefit 😀
> 
> The other thing is it seem like your issue is more easily and consistently reproducible, so hopefully someone else could set up your setup and see if they can reproduce this issue or not. Because I suppose at some point, whoever created this feature of usb virtual ethernet (including adding it to the wiki) was able to test and confirm that's it's working including @mdevaev so I wonder why it doesn't work for you in webrtc, or was it never tested in webrtc? 🤔
> 
> @Majestic7979 Frankly I don't mind volunteering to see if I can setup the same configuration (if it's not too difficult) as you and see if I can reproduce it. Are you able to join us in the Discord group?

It's literally just using the above link to enable the option... you can just copy my code for the override.yaml and add to yours, provided your override.yaml does not have extra CD/Flash drives you should have enough "lanes" to have keyboard, mouse and usb ethernet. Then plug the USB cable into the device you want to control, and you'll see it will get a new ethernet connection. Just disable the wifi on that device and any other ethernet connections, leaving only the pikvm virtual ethernet over usb to provide the connectivity. Then go to a separate device to run the chrome browser and try webrtc/h264 making sure that the audio slider is 100%. The device connected to the PiKVM is running Windows 10. The device I use for Chrome is running Windows 11. That's basically the setup. This issue seems to occur only when your browser device and the device being controlled are on separate subnets.

To clarify the network:
Device used with Chrome (to view KVM): 10.0.0.0
PiKVM itself: 10.0.0.0
Device connected to PiKVM and getting virtual ethernet from PiKVM over USB: 10.0.1.0

**Comment by plia7** [comment]
> > > > > @plia7 are you using https://docs.pikvm.org/usb_ethernet/ on your pikvm?
> 
> > > 
> 
> > > 
> 
> > > > 
> 
> > > 
> 
> > > 
> 
> > > > @RSATom No, I'm using the physical ethernet port and it works fine. Should we move @Majestic7979 thread/comments to a new ticket given it's only caused by his usb virtual ethernet (and it works fine for him otherwise)? Since it's not really related at this point to the original issue?
> 
> > > 
> 
> > > 
> 
> > > > 
> 
> > > 
> 
> > > 
> 
> > > > Suggestion: Use his existing ticket #1170 and rename the subject to "janus.js on Google Chrome black screen, or streaming cuts off periodically **when using usb virtual ethernet in local network**"
> 
> > > 
> 
> > > 
> 
> > > > 
> 
> > > 
> 
> > > 
> 
> > > > @RSATom The issue seem like the same symptoms but looks like two different root causes given the setup difference to get these symptoms. Hopefully resolving one of them should shed some light on the other one.
> 
> > > 
> 
> > > 
> 
> > > > 
> 
> > > 
> 
> > > 
> 
> > > > Thanks.
> 
> > > 
> 
> > > 
> 
> > > I am not against your idea, I hope it does not mean my issue gets less attention... Because this does seem related to a network route problem/NAT problem in both instances... whatever is decided I just ask that my issue is not forgotten, thank you 🙏🏻
> 
> > 
> 
> > Let's let @RSATom decide if it's related or not, in both cases I suppose it should work, so if my ticket gets more attention and helps to resolve yours and it leads at the end to some enlightenment with regards to the original issue I don't mind for you to keep using this thread as we both benefit 😀
> 
> > 
> 
> > The other thing is it seem like your issue is more easily and consistently reproducible, so hopefully someone else could set up your setup and see if they can reproduce this issue or not. Because I suppose at some point, whoever created this feature of usb virtual ethernet (including adding it to the wiki) was able to test and confirm that's it's working including @mdevaev so I wonder why it doesn't work for you in webrtc, or was it never tested in webrtc? 🤔
> 
> > 
> 
> > @Majestic7979 Frankly I don't mind volunteering to see if I can setup the same configuration (if it's not too difficult) as you and see if I can reproduce it. Are you able to join us in the Discord group?
> 
> 
> 
> It's literally just using the above link to enable the option... you can just copy my code for the override.yaml and add to yours, provided your override.yaml does not have extra CD/Flash drives you should have enough "lanes" to have keyboard, mouse and usb ethernet. Then plug the USB cable into the device you want to control, and you'll see it will get a new ethernet connection. Just disable the wifi on that device and any other ethernet connections, leaving only the pikvm virtual ethernet over usb to provide the connectivity. Then go to a separate device to run the chrome browser and try webrtc/h264 making sure that the audio slider is 100%. The device connected to the PiKVM is running Windows 10. The device I use for Chrome is running Windows 11. That's basically the setup. This issue seems to occur only when your browser device and the device being controlled are on separate subnets.
> 
> 
> 
> To clarify the network:
> 
> Device used with Chrome (to view KVM): 10.0.0.0
> 
> PiKVM itself: 10.0.0.0
> 
> Device connected to PiKVM and getting virtual ethernet from PiKVM over USB: 10.0.1.0

How can you see the subnets and how can you control them so they are on different subnets?

How do you access the pikvm from another device, what url do you type and how do you find what's that url?

**Comment by RSATom** [comment]
@Majestic7979 I have an idea about possible reason of your issue... Can you please tell what Raspberry Pi version your pikvm device based on?

**Comment by Majestic7979** [comment]
I will try to explain.

My router: 10.0.0.1
My PiKVM: 10.0.0.225
My PC (with Chrome): 10.0.0.16
Remote device (which the PiKVM controls): 10.0.1.3

Router gives IP to my PC running Chrome. Router gives IP to the PiKVM. Both of these devices are therefore on the same subnet (10.0.0.x).

PiKVM is connected to the Remote device via USB, that is how we all get keyboard and mouse. You can also enable a ethernet passtrough, if you do this it will create a USB-to-Ethernet bridge internally in the PiKVM, this can be enabled by the stanza in override.yaml that I posted above, specifically this part:

```
otg:
    devices:
        ethernet:
            enabled: true
            driver: ncm
otgnet:
    firewall:
        allow_tcp: [80, 443]
        forward_iface: eth0
    commands:
        post_start_cmd_append:
            - "--dhcp-option=6,8.8.8.8,8.8.4.4"
    iface:
        ip_cmd:
            - /usr/bin/ip
        net: 10.0.1.0/28
```


The bridge gives the Remote device a ethernet connection, with a subnet of 10.0.1.x, therefore the IP is 10.0.1.3 normally.

On the documentation the example is:
```
 otgnet:
    iface:
        net: 10.65.0.0/28
```
But it does not matter. You only have to make sure the subnet is different. I cannot do this:
```
 otgnet:
    iface:
        net: 10.0.0.0/28
```
because it is the same subnet of my router. The documentation makes it clear that this cannot happen. So I changed mine from 10.65.0.0 to 10.0.1.0. It does not matter what it is, as long as it's not the same as the subnet being used by the PiKVM from the router (10.0.0.0).


You should read the https://docs.pikvm.org/usb_ethernet/ in full to understand this process.

**Comment by Majestic7979** [comment]
> @Majestic7979 I have an idea about possible reason of your issue... Can you please tell what Raspberry Pi version your pikvm device based on?

My PiKVM is a Raspberry Pi 4 with the official V3 HAT, specifically it is this one here: https://shop.pimoroni.com/products/pikvm-full-kit?variant=40006251479123

This kit contains the steel case, the PiKVM V3 HAT and a Raspberry Pi 4 2GB RAM.

**Comment by RSATom** [comment]
@Majestic7979 do you have your server connected with usb 2.0 or 3.0?

**Comment by Majestic7979** [comment]
> @Majestic7979 do you have your server connected with usb 2.0 or 3.0?

It will be whatever the USB cable coming out of the PiKVM is... the server can support USB 3 because it's a USB-C/Thunderbolt port. The Windows networking on the server is showing this info for the USB ethernet:
![image](https://github.com/pikvm/pikvm/assets/128995501/8eaca0c9-531b-415c-a251-68dca31d4776)
I don't know how the implementation was done in the PiKVM for this feature so I don't know if USB speed makes a difference but we all know USB 2.0 is capped at 480Mbps so seems like the speed is USB 2.0, as we know there's keyboard and mouse also being transmitted over that connection. As you have coding experience are you able to locate and check the PiKVM source code to see how it's programmed? Maybe just to confirm if there is a hard coded limit... because if the usb ethernet was coded to only provide a max of USB 2 speed then that's all I'll ever get.

Having said the above stuff I don't know if the port from the PiKVM V3 hat is providing USB3 or USB2 (I'm talking about the port near the HDMI and ATX ports on the V3 HAT), which is USB-C and provides the emulated keyboard and mouse to the server, and also the usb ethernet if enabled. All the functions are provided by that one single connection.

**Comment by blackandcold** [comment]
Hi guys, I do not have much time (job, child, new home) to help but reading your findings.

1. Observation:

I saw the drops **only in Chromium based browsers, namely Chrome and Vivaldi. Safari works flawless and Firefox also very well**. Even after Mac sleep the KVM instantly reconnect the session.

Quality settings do not change

2. TCP/IP comment:

No USB dongle and 3 PiKVMs running with the same behaviour in the same network or remote over Tailscale. 
Most certainly there is some UDP shananigans in the Chrome network code causing the image to instantly go black. 

Bitrate and framerate most of the time show activity / traffic.
Mouse input is forwarded and when having two sessions open (Vivaldi, Safari) I can watch the mouse move on the Server in the Safari session.

My own subnet:
I have PiKVM <=> 1GBit unmanaged Switch <=Wifi AP=> M1 Mac
or PiKVM <=> 1GBit built in Switch in the AP <=> PC (EndeavourOS or Windows)

Remote:
or PiKVM <=> Tailscale <=> PC / Mac

3. Conclusion
The problem seems independent of network topology, OS and IP subnets. 

Probably leading janus to terminate the stream:
 - The drawing of the frames is not happening 
 - only metadata is transferred, images are blocked
 - Chrome code just does not forward all UDP to the active session (in time)

4. Gear:
All PiKVMs are V3 HAT,  Pi 4 4G with HDMI to CSI-2 board (Geekworm KVM-A3 Kit for Raspberry Pi 4 on Amazon).


I am a developer so hit me with janus debugging and stack traces. I'd change the JS to some containing code, setting some breakpoints and inspecting the stack before the termination.

Left Safari, right Vivaldi :)

https://github.com/pikvm/pikvm/assets/141038/16d1976c-73e5-474e-be4d-fa40031b9469

**Comment by plia7** [comment]
> Hi guys, I do not have much time (job, child, new home) to help but reading your findings.
> 
> 1. Observation:
> 
> I saw the drops **only in Chromium based browsers, namely Chrome and Vivaldi. Safari works flawless and Firefox also very well**. Even after Mac sleep the KVM instantly reconnect the session.
> 
> Quality settings do not change
> 
> 2. TCP/IP comment:
> 
> No USB dongle and 3 PiKVMs running with the same behaviour in the same network or remote over Tailscale. Most certainly there is some UDP shananigans in the Chrome network code causing the image to instantly go black.
> 
> Bitrate and framerate most of the time show activity / traffic. Mouse input is forwarded and when having two sessions open (Vivaldi, Safari) I can watch the mouse move on the Server in the Safari session.
> 
> My own subnet: I have PiKVM <=> 1GBit unmanaged Switch <=Wifi AP=> M1 Mac or PiKVM <=> 1GBit built in Switch in the AP <=> PC (EndeavourOS or Windows)
> 
> Remote: or PiKVM <=> Tailscale <=> PC / Mac
> 
> 3. Conclusion
>    The problem seems independent of network topology, OS and IP subnets.
> 
> Probably leading janus to terminate the stream:
> 
> * The drawing of the frames is not happening
> * only metadata is transferred, images are blocked
> * Chrome code just does not forward all UDP to the active session (in time)
> 
> 4. Gear:
>    All PiKVMs are V3 HAT,  Pi 4 4G with HDMI to CSI-2 board (Geekworm KVM-A3 Kit for Raspberry Pi 4 on Amazon).
> 
> I am a developer so hit me with janus debugging and stack traces. I'd change the JS to some containing code, setting some breakpoints and inspecting the stack before the termination.
> 
> Left Safari, right Vivaldi :)
> 
>  Safari.-.Vivaldi.mov

Great, I hope this helps PiKVM dev team to progress and resolve this issue.
In the meanwhile, I also posted this issue on Chromium Bugs Tracker website:
https://issues.chromium.org/issues/328677967
And Janus github issues website:
https://github.com/meetecho/janus-gateway/issues/3341

Thank you.

**Comment by RSATom** [comment]
@blackandcold is it possible for you to create WebRTC-Internals dump when issue is happens?
How long it takes to get issue after connect to pikvm from Chrome?

**Comment by Majestic7979** [comment]
I have a good feeling about this... thanks guys!

**Comment by plia7** [comment]
> I have a good feeling about this... thanks guys!

@RSATom @mdevaev The replies on the respective tickets, in janus:
"Iminiero • 5h Member "noticed that janus.js (on developer tools) keeps saying that ICE disconnected" - Then this is the cause. Not an issue in Janus, IMHO."

In Chromium:
"As per comment#1, issue seems out of scope for TE to triage which requires pikvm from an external network using tailscale installed on the external network connecting machine, hence adding TE-NeedsTriageHelp and requesting respective dev team to take a look and provide further inputs. Thanks..!!"

I guess we'll need to add some substance to these tickets that shows it's either janus.js or chromium issue and not a pikvm issue.
@RSATom Were you able to collect the debugging info from the discord war room group or they not able to reproduce on demand like Shams or not the same issue as "invalid peer connection" but different error? In that case, we might need some help with other users who have the exact issue described in this ticket.

Thank you.

**Comment by RSATom** [comment]
@mdevaev I found why Firefox behaves better:
![image](https://github.com/pikvm/pikvm/assets/1049634/8c49c357-69cf-4311-a8b9-cc1f6d165cdc)

i.e. it doesn't generate `muted` event for track in `disconnected` state. With current implementation pikvm starts reconnect to Janus and it gives `black screen` issue...

**Comment by Majestic7979** [comment]
@RSATom is this an easy fix? If so do you have an estimate how long long it could take to fix it?

**Comment by mdevaev** [comment]
@Majestic7979 It's not so easy to fix. First, we need to understand which behavior is correct.

**Comment by RSATom** [comment]
@Majestic7979 for you case I think the problem is in USB throughput through linux core. And since UDP is treated as unreliable it the first thing linux core drops in that case.

**Comment by plia7** [comment]
> @Majestic7979 It's not so easy to fix. First, we need to understand which behavior is correct.

@RSATom This is what janus dev said about this: "[lminiero](https://github.com/lminiero) commented [2 minutes ago](https://github.com/meetecho/janus-gateway/issues/3341#issuecomment-1987341707)
If you're reacting to muted events by removing the video element (or the video stream from the element), just stop doing that and Chrome will behave like Firefox."

**Comment by RSATom** [comment]
> As a short term solution (until we understand the correct behavior for the long term solution), maybe we can add this as an experimental option/flag

The problem is right now reconnect logic relies on current behavior (and btw, it looks like reconnect is broken in Firefox), and it's required to find other trigger to start Peer Connection reconnect. 
https://janus.discourse.group/t/the-latest-chrome-doesnt-assign-failed-to-iceconnectionstate/910/2

**Comment by mdevaev** [comment]
I have built the patches into a release, please update pikvm and check if the problem remains.

**Comment by RSATom** [comment]
please keep in mind, mentioned patch will not improve your network...

**Comment by Majestic7979** [comment]
> @Majestic7979 you can try reduce max fps (in pikvm settings) and screen resolution (on your server) - it should help a little bit. @mdevaev do you know, if it's possible to do the same passthrough etherenet connection but with USB 3.0 on RPi 4 ? @Majestic7979 one more idea (but I didn't try it myself, and which maybe will require some work/time on configuration) - you can try to put USB Ethernet adapter to your RPi, configure your pikvm as router, and maybe it will give you additional throughput...

I reduced the sliders and it's slightly better for sure. If it was possible to do via USB3 I would be happy to test and let you know the result.

As for configuring the PiKVM itself as a router, I'm afraid this will mess up the complicated networking situation even further? How does that work? So my ethernet cable goes into the PiKVM and then...? I am sorry, I'm not understanding what you mean by using it as a router... could you explain more please?

**Comment by mdevaev** [comment]
@Majestic7979 The audio is compressed and consumes very little traffic. You can reduce the H.264 bitrate to a minimum, maybe this will help.

@RSATom USB 3.0 is not supported by USB core hardware on Pi, so it's not possible at all.

**Comment by RSATom** [comment]
@Majestic7979 your pikvm already acts as a router - it routes traffic from virtual Ethernet adapter to physical ethernet adapter. So the only difference is use physical usb to ethernet adapter attached to USB on RPi... 

> USB 3.0 is not supported by USB core hardware on Pi, so it's not possible at all.

Oops...

**Comment by Majestic7979** [comment]
Don't worry guys, I've been working for an hour now without any irritating black screen issues, so fingers crossed for good luck that it has resolved most of the issue. Guess what, after setting up like this:
![image](https://github.com/pikvm/pikvm/assets/128995501/7a1ea45e-840d-4e52-9577-ac8f2d0e713d)

My mouse is pretty much in sync, very little lag... Pull my legs I can't believe it haha

 🍻🍻🍻🍻 Sending a box of virtual beer to the best guys on Github 😎

**Comment by Majestic7979** [comment]
Is it possible to set up the volume to be 100% as default? Every time I open a browser window I have to go into the menu and drag the slider from 0 to 100.. not an issue but would be amazing if it was possible to set the default volume as 100% :-)

**Comment by RSATom** [comment]
@Majestic7979 it's better increase gop to max - it actually reduces traffic amount....

also, FYI https://forums.raspberrypi.com/viewtopic.php?t=317326

**Comment by Majestic7979** [comment]
> @Majestic7979 it's better increase gop to max - it actually reduces traffic amount....
> 
> also, FYI https://forums.raspberrypi.com/viewtopic.php?t=317326

Thanks I increased it per your suggestion. I thought reducing the GOP would give more responsiveness as it's having to compute new frame for every frame..

Anyway this is perfect now, I have to accept that it's as good as it gets for now. Only thing that remains is setting the volume to a default of 100% so I don't have to ever open that menu again LOL. I have a remaining issue with the keyboard but I'll open a separate issue later for that.

**Comment by mdevaev** [comment]
Has everyone's problem been solved now?

**Comment by blackandcold** [comment]
Chrome works now. Vivaldi showing that is recives data, just not showing the image. Web Sockets open.
I don't know why Vivaldi does not show the content of the element but this is a different problem.
Thanks!

**Comment by mdevaev** [comment]
Perfect. So I'm finally closing it.

@RSATom thanks again for your research and patches.
